PR: https://github.com/odoo/odoo/pull/270071

From: b9065ca1561cc7054099d1177b26655164714119
From: Corentin Deruyck
Date: 2026-06-30 16:27:21

Breaking data model changes scores: del:1 + add:0, change matches:
-    require_payment = fields.Boolean(

Total Changes: 60

[IMP] sale,*: allow fixed amount for prepayment

Previously, we could only specify a down payment as a percentage.

Now users can specify a fixed amount for the down payment required
before confirming the order.

When the amount changes, the percentage is automatically recalculated,
and vice versa.

When configuring the template, only a percentage can be set.
We removed the online payment toggle and field. So now, to disable the
prepayment feature, we have to set the down payment to 0.

The require_payment field has been removed, and we now use
prepayment_percent = 0 to indicate that require_payment = False.

task-6293915

Part-of: odoo/odoo#270071
Related: odoo/enterprise#120555
Related: odoo/upgrade#10509
Signed-off-by: Victor Feyens (vfe) <vfe@odoo.com>

================================= pseudo patch: =================================

--- a/addons/sale/models/res_company.py
+++ b/addons/sale/models/res_company.py
@@ -19,7 +19,7 @@ class ResCompany(models.Model):
     portal_confirmation_sign = fields.Boolean(string="Online Signature", default=True)
     portal_confirmation_pay = fields.Boolean(string="Online Payment")
     prepayment_percent = fields.Float(
-        string="Prepayment percentage",
+        string="Prepayment",
         default=1.0,
         help="The percentage of the amount needed to be paid to confirm quotations.",
     )
@@ -76,7 +76,7 @@ class ResCompany(models.Model):
     @api.constrains("prepayment_percent")
     def _check_prepayment_percent(self):
         for company in self:
-            if company.portal_confirmation_pay and not (0 < company.prepayment_percent <= 1.0):
+            if company.portal_confirmation_pay and not (0 <= company.prepayment_percent <= 1.0):
                 raise ValidationError(
                     company.env._("Prepayment percentage must be a valid percentage.")
                 )

--- a/addons/sale/models/sale_order.py
+++ b/addons/sale/models/sale_order.py
@@ -142,22 +142,22 @@ class SaleOrder(models.Model):
         precompute=True,
         help="Request a online signature from the customer to confirm the order.",
     )
-    require_payment = fields.Boolean(
-        string="Online payment",
-        compute="_compute_require_payment",
-        store=True,
-        readonly=False,
-        precompute=True,
-        help="Request a online payment from the customer to confirm the order.",
-    )
     prepayment_percent = fields.Float(
-        string="Prepayment percentage",
+        string="Prepayment",
         compute="_compute_prepayment_percent",
         store=True,
         readonly=False,
         precompute=True,
         help="The percentage of the amount that must be paid by the customer to confirm the order.",
     )
+    prepayment_amount = fields.Monetary(
+        string="Prepayment Amount",
+        help="The amount that must be paid by the customer to confirm the order.",
+        currency_field="currency_id",
+        compute="_compute_prepayment_amount",
+        inverse="_inverse_prepayment_amount",
+        readonly=False,
+    )
 
     signature = fields.Image(
         string="Signature", copy=False, attachment=True, max_width=1024, max_height=1024
@@ -521,14 +521,24 @@ class SaleOrder(models.Model):
             order.require_signature = order.company_id.portal_confirmation_sign
 
     @api.depends("company_id")
-    def _compute_require_payment(self):
+    def _compute_prepayment_percent(self):
         for order in self:
-            order.require_payment = order.company_id.portal_confirmation_pay
+            if order.company_id.portal_confirmation_pay:
+                order.prepayment_percent = order.company_id.prepayment_percent
+            else:
+                order.prepayment_percent = 0.0
 
-    @api.depends("require_payment")
-    def _compute_prepayment_percent(self):
+    @api.depends("prepayment_percent", "amount_total")
+    def _compute_prepayment_amount(self):
+        for order in self:
+            order.prepayment_amount = order.prepayment_percent * order.amount_total
+
+    def _inverse_prepayment_amount(self):
         for order in self:
-            order.prepayment_percent = order.company_id.prepayment_percent
+            if order.amount_total:
+                order.prepayment_percent = order.prepayment_amount / order.amount_total
+            else:
+                order.prepayment_percent = 0.0
 
     @api.depends("company_id")
     def _compute_validity_date(self):
@@ -1226,7 +1236,7 @@ class SaleOrder(models.Model):
     @api.constrains("prepayment_percent")
     def _check_prepayment_percent(self):
         for order in self:
-            if order.require_payment and not (0 < order.prepayment_percent <= 1.0):
+            if not (0 <= order.prepayment_percent <= 1.0):
                 raise ValidationError(
                     order.env._("Prepayment percentage must be a valid percentage.")
                 )
@@ -1303,10 +1313,10 @@ class SaleOrder(models.Model):
             self.order_line and self._origin.pricelist_id != self.pricelist_id
         )
 
-    @api.onchange("prepayment_percent")
-    def _onchange_prepayment_percent(self):
-        if not self.prepayment_percent:
-            self.require_payment = False
+    @api.onchange("prepayment_amount")
+    def _onchange_prepayment_amount(self):
+        if self.amount_total:
+            self.prepayment_percent = self.prepayment_amount / self.amount_total
 
     @api.onchange("order_line")
     def _onchange_order_line(self):
@@ -2306,7 +2316,7 @@ class SaleOrder(models.Model):
 
         prepayment_amount = self._get_prepayment_required_amount()
         remaining_balance = self.amount_total - self.amount_paid
-        if self.state in ("draft", "sent") and self.require_payment:
+        if self.state in ("draft", "sent") and self.prepayment_percent > 0:
             suggested_amount = prepayment_amount  # Suggest the amount needed to confirm the quote.
         else:  # The order is confirmed or doesn't require payment.
             suggested_amount = remaining_balance
@@ -2371,7 +2381,7 @@ class SaleOrder(models.Model):
         A sale order has to be paid when:
         - its state is 'draft' or `sent`;
         - it's not expired;
-        - it requires a payment;
+        - the prepayment percent is strictly positive;
         - the last transaction's state isn't `done`;
         - the total amount is strictly positive.
         - confirmation amount is not reached
@@ -2385,7 +2395,7 @@ class SaleOrder(models.Model):
         return (
             self.state in ["draft", "sent"]
             and not self.is_expired
-            and self.require_payment
+            and self.prepayment_percent > 0
             and self.amount_total > 0
             and not self._is_confirmation_amount_reached()
         )
@@ -2580,7 +2590,7 @@ class SaleOrder(models.Model):
         """
         self.ensure_one()
 
-        if not self.require_payment:
+        if self.prepayment_percent == 0:
             return 0
         return self.currency_id.round(self.amount_total * self.prepayment_percent)
 
